### PR TITLE
Closes #170  - implement event notification criteria

### LINKED ISSUE #170: Implement notification criteria
https://docs.google.com/document/d/101jsPjZZ-a_MgZn1UaAyJouufkxzJqG07IrI_0USj1M/edit#heading=h.orhdwaxry889

This criteria determines:

1. When we should start looking for events
2. When general (non-Mozilla) events should be notified
3. When Mozilla events should be notified

### PR DESCRIPTION
**Pull request by fluffyemily**
There is one question open to Bram at the moment. See comment on issue #170

### CODE REVIEW

**Comment by fluffyemily** [comment]
[I want to receive accurate and timely notifications of upcoming events near me, so that I can be pleasantly surprised while exploring the city](https://trello.com/c/UpzdyKrc/50-i-want-to-receive-accurate-and-timely-notifications-of-upcoming-events-near-me-so-that-i-can-be-pleasantly-surprised-while-explo)

File: Prox/Prox/Data/EventsController.swift [hunk 1]
```diff
@@ -47,25 +51,52 @@ class EventsProvider {
...
         return event.startTime <= maxStartTime
     }
 
     private func isEventToday(event: Event) -> Bool {
         return Calendar.current.isDateInToday(event.startTime)
     }
 
-    private func isEventYetToHappen(event: Event) -> Bool {
-        return Date() < event.startTime
+    private func isFutureEvent(event: Event, currentTime: Date) -> Bool {
```
Comment (mcomella) [inline Prox/Prox/Data/EventsController.swift]: nit: -> `hasEventStarted`

File: Prox/Prox/Data/EventsController.swift [hunk 2]
```diff
@@ -47,25 +51,52 @@ class EventsProvider {
...
+    private func isValidFutureEvent(event: Event, currentTime: Date, forNotifications: Bool = false) -> Bool {
+        let startTimeInterval = forNotifications ? eventStartNotificationInterval : eventStartPlaceInterval
+        return isEventToday(event: event) && isFutureEvent(event: event, currentTime: currentTime) && doesEvent(event: event, startAtTimeInterval: startTimeInterval, fromTime: currentTime)
+    }
+
+    private func isValidCurrentEvent(event: Event, currentTime: Date, forNotifications: Bool = false) -> Bool {
+        return isCurrentEvent(event: event, currentTime: currentTime) && doesEvent(event: event, lastLessThan: maxEventDuration)
+    }
+
+    private func doesEvent(event: Event, startAtTimeInterval timeInterval: TimeInterval, fromTime time: Date) -> Bool {
```
Comment (mcomella) [inline Prox/Prox/Data/EventsController.swift]: nit: -> `doesEvent(..., startInTimeIntereval, ...)`

"At" indicates a single moment in time to me so `startAtTimeInterval` makes me think you're checking exact times.

File: Prox/Prox/Utilities/EventNotificationsManager.swift [hunk 3]
```diff
@@ -13,13 +13,13 @@ let notificationEventIDKey = "eventPlaceID"
 class EventNotificationsManager {
 
     // caches events by their start time by place
-    fileprivate lazy var sentNotifications: [String: Set<Date>] = {
-        var cache = [String: Set<Date>]()
-        guard let savedNotifications = UserDefaults.standard.dictionary(forKey: sentNotificationDictKey) as? [String: Set<Date>] else {
+    fileprivate lazy var sentNotifications: [String: [Double]] = {
```
Comment (mcomella) [inline Prox/Prox/Utilities/EventNotificationsManager.swift]: nit: What is this a mapping between? `sentNotifications` as a name doesn't tell me much.

File: Prox/Prox/Data/EventsController.swift [hunk 4]
```diff
@@ -62,11 +66,19 @@ class EventsProvider {
...
     }
 
     private func isValidCurrentEvent(event: Event, currentTime: Date, forNotifications: Bool = false) -> Bool {
-        return isCurrentEvent(event: event, currentTime: currentTime) && doesEvent(event: event, lastLessThan: maxEventDuration)
+        guard isCurrentEvent(event: event, currentTime: currentTime), let endTime = event.endTime else { return false }
+
+        return doesEvent(event: event, lastLessThan: maxEventDuration) || isTime(time: endTime, withinTimeInterval: minTimeFromEndOfEventForNotificationMins, fromTime: currentTime)
+    }
+
+    private func isTime(time: Date, withinTimeInterval timeInterval: TimeInterval, fromTime startTime: Date) -> Bool {
```
Comment (mcomella) [inline Prox/Prox/Data/EventsController.swift]: nit: `doesEvent(..., startAtTimeInterval, ...)` should call out to this.

**Pull request merged**